Enhanced parsing of input values (post or parameters)

jlannoy opened the pull request:
The goal of this PR is to enhance the parsing of strings to simple types, from the parameters or from the request. 
In details, it allows :
- To avoid duplicated code or differences between ParamParsers and SwissKnife converters (their job are in fact the same)
- To allow the POST body engine to register validation errors on conversion failure
- To avoid registering enumeration at start
- To allow users to register their own ParamParser
- Even overriding default ones, which can be useful to specify a particular DateFormat

Notes : 
- Custom ParamParsers are instancied via an Injector, so that users can make usage of @Inject dependencies
- The convert method for Date from the SwissKnife class have been translated to a ParamParser, based on LocalDateTime (as the original method). I’m thinking about something more useful, with date format from the properties file based on current Locale ; but I didn’t want to make a bigger PR. The fact is that with this PR, each user will be able to write a parser with it’s own format.

jlannoy committed:
Documentation enhanced

More information about custom configuration of the Freemarker template
engine, with small code examples.

jlannoy committed:
Enhancing the parsing of strings to simple types, from the parameters or
from the request

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy commented:
Raphael, can you do a first review of this? If OK, then I'll check what's going on with the branch conflicts (it must be because I started it before last release).

jlannoy committed:
NinjaProperties in main module

Allow the main module to access the NinjaProperties if needed. Backward
compatible.

raphaelbauer commented:
Hey! This looks good (I just roughly looked at the code). Some really good stuff in this PR! Could you make CI pass so that I can run the code? Thanks!

jlannoy committed:
Exception message enhanced

jlannoy committed:
Synchronizing changelog

jlannoy committed:
Moving from SwissKnife to ParamParsers

No more parsing operation in the SwissKnife class.

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

jlannoy commented:
Voila. Now it's okay ;-)

raphaelbauer commented:
Checking out the PR now... :)

raphaelbauer commented on ninja-core/src/main/java/ninja/utils/SwissKnife.java:
```diff
@@ -40,19 +34,6 @@
 
     private static final Logger logger = LoggerFactory.getLogger(SwissKnife.class);
 
-    private static final Map<String, Method> CONVERTERS = new HashMap<>();
```
Cleaning SwissKnife up and having that code only once is genius... :+1:

raphaelbauer commented on ninja-core/src/site/markdown/documentation/basic_concepts/controllers.md:
```diff
@@ -125,7 +125,10 @@ If your custom object has a date/timestamp field, first make sure that you use t
-accepted by Joda Time library.
+accepted by Joda Time library. Or you can register a <code>ninja.params.ParamParser</code> with your own date format parsing in your <code>conf.Module</code>:
+<pre class="prettyprint">
+ParamParsers.registerParamParser(Date.class, MyDateParamParser.class);
```
That makes me think whether it is clever that ParamParsers is static. (Not an issue of this PR). Do you think we should create an issue for that?

jlannoy replied:
Not sure. I think it's quite the same discussion as the NinjaProperties given to a module as constructor arg... If it is not static, you'll not be able too to inject it in the conf.Module. So where will users do the early job of registering parsers? 

Not that, with the way I did the PR, at least all custom parsers will now be Guice-aware.

raphaelbauer replied:
(It's not meant to be a critique of your PR). I added my "solution" to the discussion page. Would be great getting feedback on that...

raphaelbauer commented on ninja-core/src/main/java/ninja/bodyparser/BodyParserEnginePost.java:
```diff
@@ -62,21 +75,34 @@
+                    
+                    if (Collection.class.isAssignableFrom(fieldType) || List.class.isAssignableFrom(fieldType)) {
+                    	
+                    	ListParamParser<?> parser = (ListParamParser<?>) ParamParsers.getListParser(getGenericType(field));
```
I guess getListParser should get the injector... (Is that something unit tests should have caught?)

jlannoy replied:
You are right. I made a unit test for a user custom Date parser that needs a injector, but I forget to test Array and List parsers that would use that custom parser. I'll fix it tomorrow.

raphaelbauer commented on ninja-core/src/main/java/ninja/bodyparser/BodyParserEnginePost.java:
```diff
@@ -37,6 +43,13 @@
 
     private final Logger logger = LoggerFactory.getLogger(BodyParserEnginePost.class);
 
+	Injector injector;
```
this can be private

raphaelbauer commented on ninja-core/src/main/java/ninja/params/ParamParsers.java:
```diff
@@ -53,61 +60,95 @@
+    /**
+     * ALlows to register a new parser for given Class parameters.
+     */
+    public static <T> void registerParamParser(final Class<T> paramClass, final Class<? extends ParamParser<T>> parserClass) {
```
I'd like to have a big warning sign that registerParamParser must only be used at startup of Ninja. All other usages will not work in a multi threaded context with PARAM_PARSERS_CLASSES's HashMap. The same should be added for unregisterParamParsers.

Maybe that is another sign to rewrite that class to a non static class (Singleton). (I guess not part of this PR, just saying).

raphaelbauer commented on ninja-core/src/main/java/ninja/params/ParamParsers.java:
```diff
@@ -53,61 +60,95 @@
-            return ENUM_PARSERS.get(targetType);
+        
+        if (targetType.isEnum()) {
+        	return new GenericEnumParamParser((Class<Enum>) targetType);
```
there's a hidden tab :) should be 4 spaced ;)

raphaelbauer commented:
Hey @jlannoy , PR looks good. Very thoughtful and well done. The only thing that I have in my mind is rewriting ParamParsers from being static to being a @Singleton.

Injection of ParamParsers can then be done via https://github.com/google/guice/wiki/Multibindings . The additional benefit is that we can make sure that no user violates the contract of register / unregisterParamParser.

The disadvantage is that adding custom Param parsers is a bit more "complex":

```
   Multibinder<ParamParser> uriBinder = Multibinder.newSetBinder(binder(), ParamParser.class);
   uriBinder.addBinding().toInstance(new MyFunkyParamParser);
```

That's not something that you have to do I guess... I got some time and can do it on top of your PR if that is ok with you... Let me know what you think. Maybe I missed something...

raphaelbauer commented on ninja-core/src/test/java/ninja/bodyparser/BodyParserEnginePostTest.java:
```diff
@@ -143,4 +293,15 @@ public void testBodyParserWithCollectionAndArray() {
     }
     
+    public static enum MyEnum {
+    	VALUE_A, VALUE_B, VALUE_C
```
there is a nasty tab hidden in the code ;)

jlannoy commented:
No problem, I completely agree on your comments. I didin't had a better solution in mind when I made the PR ; so I copied the static registerEnum() way of working. I'll check tomorrow for the MultiBindings. I'll also fix the other notes you made during your review.

jjlauer commented:
This is a cool PR.  Looking fwd to it being integrated. +1 on non-static /
multibinding solution if it works.

On Thu, Jan 28, 2016 at 6:40 AM, J Lannoy notifications@github.com wrote:

> No problem, I completely agree on your comments. I didin't had a better
> solution in mind when I made the PR ; so I copied the static registerEnum()
> way of working. I'll check tomorrow for the MultiBindings. I'll also fix
> the other notes you made during your review.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ninjaframework/ninja/pull/448#issuecomment-176138241.

raphaelbauer committed:
Final cleanup. Module.java can now get NinjaProperties injected.

raphaelbauer committed:
Fix Cookie domain is not set when clearing session #462

jlannoy committed:
Tabs fixed and more tests added

All hidden tabs have been replaced by spaces and more tests have been
added to check the List and Array parsing, and to check the injection
inside a custom parser.

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy committed:
Merge branch 'develop' into feature-bodyparser-improvement

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy committed:
Changelog updated

jlannoy committed:
Merge pull request #2 from jlannoy/feature-bodyparser-improvement

Merge to 5.3.1

jlannoy committed:
Merge branch 'develop' into feature-bodyparser-improvement

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

jlannoy committed:
ParamParsers as a singleton

Guice instancied ; custom parsers are inject through a multibinder.

jlannoy committed:
ParamParsers as a singleton

Guice instancied ; custom parsers are inject through a multibinder.

jlannoy committed:
Fixing tests that depends on AbstractContext

Need to pass ParamParsers

jlannoy committed:
Fixing tests that depends on AbstractContext

Need to pass ParamParsers

jfendler committed:
Added link to Ninja-Quartz module in modules.md documentation.

raphaelbauer committed:
Removed comment that was no longer up-to-date

jlannoy committed:
Imports fixed

jlannoy committed:
Imports fixed

jlannoy committed:
Documentation for MultiBinder

jlannoy committed:
Documentation for MultiBinder

jlannoy committed:
Context default values

jlannoy committed:
Context default values

jlannoy commented:
Hi @raphaelbauer and @jjlauer.
I finally managed to do it with a _Multibinder<ParamParser>_. I also add tests for the collections and array parsers. 

But... It's not the end. I've discovered that the way I did it presents a conceptual error. Reusing the ParamParsers in the BodyParserEnginePost will allow to get parsing errors in the Validation object but.. Actually, they are treated as "FileViolations", not as "BeanViolations" as it should be, following the Validation object javadoc. So, how to solve this problem?

I can modify the ParamParser interface to add one method or one attribute to the current parsing method, allowing to know if it's for a bean of a field... Or we can rethink a little bit the Validation object. Is it really needed to separate the two? ("Field" is a little bit inappropriate (personal opinion), it should for example be "Parameter".) It's an open question, please give my your advice on the best way to handle that.

Jonathan

jjlauer committed:
Update docs

jfendler committed:
Added link to Ninja Sitemap module to documentation.

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy committed:
Changelog

jlannoy committed:
Changelog

jlannoy committed:
Changelog

jjlauer committed:
Added RecycledNinjaServerTester for speedy unit tests

jjlauer committed:
Removed ninja-core dependency on org.mindrot:bcrypt (it was unused)

jjlauer committed:
More flexible guice configuration!

jjlauer committed:
New AutoStandalone class to find a standalone for running and unit tests

jjlauer committed:
Fixed import that broke travis

jjlauer committed:
Routing accepts a new MethodReference; reverse routing now MUCH faster

jlannoy committed:
Make diff with bean and field validations

jlannoy committed:
Make diff with bean and field validations

jlannoy committed:
Update from develop

jlannoy committed:
Enhance ValidationImpl performance

As validations lists are mainly used by field id, use maps to enhance
performance

jlannoy committed:
Enhance ValidationImpl performance

As validations lists are mainly used by field id, use maps to enhance
performance

jjlauer committed:
Added unit test for ImplFromPropertiesFactory

jjlauer committed:
Unit tests for MethodReference usage

jjlauer committed:
Fixed travis build

jlannoy committed:
Revert "Make diff with bean and field validations"

This reverts commit 4d72850d9763d993c7616a33a26faafb70961082.

jlannoy committed:
Revert "Make diff with bean and field validations"

This reverts commit 4d72850d9763d993c7616a33a26faafb70961082.

jlannoy committed:
Validation doc updated

So that the difference between Field and Bean validation is explained.

jlannoy committed:
Validation doc updated

So that the difference between Field and Bean validation is explained.

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature-bodyparser-improvement

# Conflicts:
#	ninja-core/src/site/markdown/developer/changelog.md

jlannoy committed:
Fix for NinjaBaseModule

jlannoy committed:
Fix for NinjaBaseModule

jlannoy commented:
Hi @raphaelbauer and @jjlauer.

It's well done now.
- All your comments have been fixed
- It works with the MultiBinder, and is documented as well
- I had to enhanced the Validation documentation page, to make the differences with the JSR303 validator clear (interpolated messages in bean violation // message keys in field validations)

Work done for me.

jjlauer commented:
Hey @jlannoy -- that's a pretty large commit.  Its a bit difficult to provide fresh feedback. Wondering if you tried/could rebase and force a fresh push to your branch so we can review your change in a single commit?  Here's one blog explaining the process a bit more: https://robots.thoughtbot.com/keeping-a-github-fork-updated

jlannoy commented:
I'll see when I can try to do it. Note that since @raphaelbauer's original review, not so much code has change. A lot of commits are due to merges with the develop branch (and conflicts one the changelog).

jlannoy committed:
Rebase conflicts resolved

jlannoy committed:
Rebase conflicts resolved

jlannoy committed:
Merge branch 'feature-bodyparser-improvement' of https://github.com/jlannoy/ninja into feature-bodyparser-improvement

jlannoy commented:
(Rewrite modifications in a new branch)
